### ISSUE TITLE
Unhandled rejection when joining chains with common rejected parent

### ISSUE THREAD

**Issue by tjconcept** [petkaantonov/bluebird#645]
Maybe I should do this kind of stuff in another way, but I've hit it a couple of times now and it's extremely hard to debug and, at least in my head, seems inconsistent and unintuitive.

```js
var Promise = require('bluebird');

var a = Promise.try(function(){
	throw new Error('Something went wrong');
});

var b = Promise.try(function(){
	throw new Error('Something went wrong here as well');
});

var c = Promise
	.join(a, b)
	.spread(function( a, b ){
		return a+b;
	});

Promise
	.join(a, c)
	.spread(function( a, product ){
		// ...
	})
	.catch(Error, function(e){
		console.log(e);
	});
```

The workaround is to "silent" the `a`-chain in the last join with `.catch(noop)` (#483 in the future), but then `c` needs to know about the fact that `a` is part of `c` which is not always easy.

Does it make sense?

**Comment by benjamingr** [comment]
Can you check if this happens with the correct .join syntax?

**Comment by tjconcept** [comment]
It does not happen with `join(a, c, fn)`, if that is what you mean?

I was not aware this was incorrect.. How come?

**Comment by benjamingr** [comment]
The old join syntax was deprecated and then superseded with the (a, b, fn) one a while ago.

That said, the face of still works (the old syntax) means that either:

 - the behavior in terms of unhandled rejection tracking should be fixed 
 - 3.0 should remove it and give a big fat warning

Either is fine by me - the current behavior isn't good.

**Comment by tjconcept** [comment]
What were the reasons for deprecating the "old" syntax? I think it aligns nicely with `.all`, `.props` and so forth. How is it different?

**Comment by spion** [comment]
You can always use Promise.all

    Promise.all([a,b]).bind(console).spread(console.log)

though when I proposed the original change for `join` I didn't really mean to suggest deprecating the old one - its just that it was much easier to model the type signatures in TypeScript for the new syntax

**Comment by tjconcept** [comment]
Hmm.. I really think of `join` and `spread` as two different things, just as `resolve` and `then`.
`spread` is calling the function with `.apply` or using the spread operator from ES6, regardless of the actual resolved value.
`join` is combining several values into an array of resolved values - like using rest parameters.

**Comment by tjconcept** [comment]
> You can always use Promise.all

Seems like `.all` is giving me the same behaviour as `.join().spread()` with an uncaught exception.

**Comment by petkaantonov** [comment]
Yea it's a bug that I fixed a while ago in 3.0 with https://github.com/petkaantonov/bluebird/commit/b182d5b8c1ee09501ea50478c64a8f206c00b304 . Maybe it could be backported

**Comment by petkaantonov** [comment]
Fixed in 2.9.28
